### ISSUE TITLE
banaction=route generates lots of errors on restarting fail2ban service

### ISSUE THREAD

**Issue by tvhdeeptho** [fail2ban/fail2ban#2373]
### Environment:

- Fail2Ban version (including any possible distribution suffixes):
-Fedora 20
- [x] Fail2Ban installed via OS/distribution mechanisms 
- [ ] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:
The "route" banaction uses "ip route add" to ban a host. However, if a route
already exists, this generates an error message which fills up the log:

2019-03-17 15:06:43,218 fail2ban.utils         [19955]: #39-Lev. 7f42e427ace8 -- exec: ip route add unreachable 157.7.129.162
2019-03-17 15:06:43,218 fail2ban.utils          [19955]: ERROR   7f42e427ace8 -- stderr: 'RTNETLINK answers: File exists'
2019-03-17 15:06:43,218 fail2ban.utils          [19955]: ERROR   7f42e427ace8 -- returned 2
2019-03-17 15:06:43,218 fail2ban.actions        [19955]: ERROR   Failed to execute ban jail 'repeat-offender' action 'route' info 'ActionInfo({'ip': '157.7.129.162', 'fid': <function Actions.ActionInfo.<lambda> at 0x7f42e60922f0>, 'raw-ticket': <function Actions.ActionInfo.<lambda> at 0x7f42e6092840>})': Error banning 157.7.129.162
2019-03-17 15:06:43,218 fail2ban.actions        [19955]: NOTICE  [repeat-offender] Restore Ban 157.7.132.37
2

#### Steps to reproduce
set banaction = route on sshd fail

systemctl restart fail2ban.service

#### Expected behavior
procduces many error messages

#### Observed behavior
may error messages

#### Any additional information
A very simple solution is to replace "ip route add" with "ip route replace"

#### Any customizations done to /etc/fail2ban/ configuration
```
```

#### Relevant parts of /var/log/fail2ban.log file:
_preferably obtained while running fail2ban with `loglevel = 4`_

```
```

#### Relevant lines from monitored log files in question:

```
```

**Comment by sebres** [comment]
Looks similar to https://github.com/fail2ban/fail2ban/issues/2031#issuecomment-361185607 for shorewall action.

> A very simple solution is to replace "ip route add" with "ip route replace"

This would be rather a workaround. 
As I understand the IP will be banned by several jails (which action route cannot differentiate at the moment). So if in jail1 the IP gets banned for 5 minutes and in jail2 for 1 hour, the replace seems to work, but if it will be unbanned from jail1 (in 5 minutes), it becomes then "unprotected" regarding jail2.

Either actions like this should get some counter or fail2ban should get common ban-manager for all-port actions without anchoring/chains/etc.

**Comment by dihedral** [comment]
I am using various metric values for each jail as to distinguish the route entries. seems to work:

action.d/route.conf
```
[Definition]
actionban   = ip route add <blocktype> <ip> <additional>
...
[Init]
additional = 
```

jail.local
```
[sshd]
enabled = true
action=%(banaction)s[additional="metric 30"]
```

**Comment by sebres** [comment]
Metric is used to make a routing decisions (so rather serves precedent purposes). Usage of metric for identification purposes sounds like a workaround either.

**Comment by dihedral** [comment]
Yes, it is a workaround. But with using metric as identifier, I do not have your described issue of potentially lifting a ban when two jails handle the same IP.

**Comment by sebres** [comment]
Sure, but it remains... ugly.
Proper solution would be either some identifier for `ip route` (derived from jail name) or an jail independent ban-manager (sharing banned ticket and banning with single action across multiple jails).
First is an action related stuff and depends on functionality of `ip route` (I have no idea it is possible somehow), second would be enhancement of fail2ban (its internal functionality) and could theoretically work for all actions.

**Comment by dihedral** [comment]
Value it as you like, this outlines a functioning workaround until the perfect solution exists.

**Comment by dihedral** [comment]
would it be possible for a ban-manager as you call it, to be an action itself and in turn trigger another action if considered necessary?

i.e. banaction = banmanager[nextbanaction=route]

and how would such a ban-manager then need to trigger <nextbanaction> from its own actionban line(s)?

**Comment by sebres** [comment]
As for ban-manager - not really.

BanManager already exists in internal fail2ban API (an object at the moment holding the banned tickets per each jail).
To provide this functionality, it should be extended to share single manager across several jail, should get ref-counter for the banned tickets (coming from 2 jails), etc.
It's not very complex (diligence work), but also not so trivial, that one can implement it (with test-cases etc) in few hours.

As for "action itself" - there are indeed already similar actions in fail2ban (executing another banning actions), like [badips](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/badips.py). 
But doing that for sharing purposes only is a bit ugly - you have to hold a duplicate list for the shared tickets (with ref-counter and prolongation of bantime), so that would be simply too much for an action.

**Comment by jucajuca** [comment]
I am experiencing the same issue, what would be the solution for this?

**Comment by sebres** [comment]
There is no ultimate solution at the moment for that (as you probably see the issue is still open)...
Only workarounds like suggested by @dihedral above (in https://github.com/fail2ban/fail2ban/issues/2373#issuecomment-668229261)
